spring-integration-core/src/main/java/org/springframework/integration/aggregator/ResequencingMessageHandler.java

```diff
@@ -81,13 +83,21 @@ protected void afterRelease(MessageGroup messageGroup, Collection<Message<?>> co
 			remove(messageGroup);
 		}
 		else {
-			if (completedMessages != null){
-				int lastReleasedSequenceNumber = this.findLastReleasedSequenceNumber(messageGroup.getGroupId(), completedMessages);
-				messageStore.setLastReleasedSequenceNumberForGroup(messageGroup.getGroupId(), lastReleasedSequenceNumber);
-				this.messageStore.removeMessagesFromGroup(messageGroup.getGroupId(), completedMessages);
+			Object groupId = messageGroup.getGroupId();
+			if (completedMessages != null) {
+				int lastReleasedSequenceNumber = findLastReleasedSequenceNumber(groupId, completedMessages);
+				this.messageStore.setLastReleasedSequenceNumberForGroup(groupId, lastReleasedSequenceNumber);
+				if (this.messageStore instanceof SimpleMessageStore
+						&& completedMessages.size() == messageGroup.size()) {
+					((SimpleMessageGroup) messageGroup).clear();
+					((SimpleMessageGroup) messageGroup).setLastModified(System.currentTimeMillis());
```

Comment:
Unfortunately, this fix isn't good enough. We end up with the same `non-released permit` issue here again with limited `MessageStore`:
> org.springframework.messaging.MessagingException: SimpleMessageStore was out of capacity at for individual group, try constructing it with a larger capacity.

I still don't see way how to fix...